### ISSUE TITLE
Add PATCH support for AJAX update

### ISSUE THREAD

**Issue by whatyoubendoing** [spine/spine#404]
Allow Spine to send PATCH requests for models, which allow partial updates of resources.

**Comment by jeremyhaile** [comment]
@aeischeid I saw you added a milestone and then removed it. Do you know if this is something that someone is already working on?

**Comment by jeremyhaile** [comment]
Are there any branches or forks that support sending partial updates?  This is really important to us, as we're having trouble with users overwriting other user's changes because Spine sends ALL fields to the server instead of only sending changed fields.  So if one user changes field A and another user changes field B...the second user might also overwrite field A, since Spine is sending the whole model to the server every time.

We're having to consider alternatives such as:

1. Trying to patch Spine model `updateAttributes` ourselves to only send the attributes that you are updating (which isn't straight-forward given the way Spine's internal updating and events work)
2. Not using Spine's AJAX model support at all and sending the updates manually ourselves.
3. Migrating off of Spine to a different JavaScript framework

**Comment by richard-flosi** [comment]
@jeremyhaile I'm not aware of any PATCH support anywhere in Spine currently. Models would need to probably start using Observers first. I'd suggest using custom ajax requests and/or work on submitting a pull request for the PATCH feature. The hard part here is knowing which fields to send and the model doesn't currently track that.

The Spine framework is fairly loosely coupled. I imagine it wouldn't be hard to use a different model system with Spine. For example Backbone models support patch by passing the `{patch: true}` option.
Docs: http://backbonejs.org/#Model-save
You could probably replace all or some of your Spine models with Backbone models as needed for now.

You also might want to consider a different architecture for models that have this problem.
Take a look at these to videos for some ideas:
Full Stack Flux: https://www.youtube.com/watch?v=KtmjkCuV-EU
Turning the Database Inside Out: https://www.youtube.com/watch?v=fU9hR3kiOK0

Good luck and let us know what you decide to do. Thanks!

**Comment by adambiggs** [comment]
Quick update. It doesn't look like Object.observe is going to be a thing after all:

https://esdiscuss.org/topic/an-update-on-object-observe

**Comment by jeremyhaile** [comment]
@adambiggs Thanks for the update. 

Whenever model attributes are updated, they flow through the `Model#load` method. It seems like we should add "dirty tracking" in `load` to keep track when attributes change.

Then when you call `Model#save` it would only send the "changed" properties to the server if patch mode is enabled.

We could add a mirror version of the `attributes` method called `changedAttributes` that only returns the changed attributes. The AJAX code currently just calls `@record.toJSON()` to get the changes to send, but would need to be modified to only get JSON for the changed attributes. This is tricky because `toJSON` confusingly doesn't return JSON but just a hash of attributes to send to the server. We could add a `toChangedJSON` that by default calls `toJSON` and then filters out all except for the "changed" attributes. This way it wouldn't disrupt anyone's custom toJSON implementation. Unless of course they were adding in additional attributes, mapping attribute names, etc. But I'm not sure how to efficiently work around that with patch mode enabled.

One factor which needs to be taken into consideration is whether or not the item is already persisted, which Spine's model doesn't really address today.  If you call `Model.create(...)` the model could be a new object or could simply be a reflection of a server-side object that already exists.  In the case of an already-existing object, you'd consider none of the fields dirty. In the case of a newly created object, you'd consider all of the fields dirty.

We could address that by either detecting the presence of an `id` as representing the fact that it's already persisted. Or we could consider `create` as creating a new object and `refresh` as representing existing objects.

Any thoughts? Other ideas on how to implement this?

**Comment by jeremyhaile** [comment]
A **much** simpler but less powerful implementation would be:

- Modify `Model.updateAttributes` to include the attributes in the options passed to `save` (for example: `options._attrs = attrs`)
- Since `save` passes those options on all the way down into the ajax code, the `ajax` module could then just examine `options._attrs` if `options.patch == true` and then only send those attributes to the server

**Comment by adambiggs** [comment]
Thanks for restarting the discussion @jeremyhaile.

> This is tricky because toJSON confusingly doesn't return JSON but just a hash of attributes to send to the server.

This is intended to take advantage of [native `JSON.stringify()` behaviour](https://developer.mozilla.org/en/docs/Web/JavaScript/Reference/Global_Objects/JSON/stringify#toJSON()_behavior). I think we'd want to keep this untouched so that `JSON.stringify(model)` always returns the same result.

> We could address that by either detecting the presence of an id as representing the fact that it's already persisted. Or we could consider create as creating a new object and refresh as representing existing objects.

The app I'm working on generates unique IDs client-side, so that can't be used to determine the server-state of the record... Currently if the record exists in model storage, it's assumed to also exist on the server.

> A much simpler but less powerful implementation...

It would be really cool if we could get this working with `save()`, `updateAttribute()`, and `updateAttributes()`. Out of these I personally use `save()` the most.

**Comment by adambiggs** [comment]
Assuming we have a way to get dirty changes (like you're suggested `changedAttributes()` method, or by passing an option into the existing `attributes()` method), maybe it could go something like this:

- In `Model::update()`, when an 'update'/'change' event is triggered, we pass 2 extra bits of data along with the options object: `options.changed_attrs`, and `options.previous_attrs`.
- When the Ajax module sees the 'change' event, a `PATCH` request is sent with `changed_attrs`.
- If the request fails, the model is refreshed with `previous_attrs` to take care of #369.

**Comment by adambiggs** [comment]
We should be able to build a `changed_attrs` object pretty easily by looking at a clone's `own` attributes before `stripOwnAttrs()` is called on it. Then we can use the keys of that object to build `previous_attrs` from model storage before applying the changes.

I don't think doing it like this would be too expensive... But it's a concern.

**Comment by jeremyhaile** [comment]
@adambiggs I like your ideas on passing `changedAttrs` and `previousAttrs`, and the idea of refreshing from `previousAttrs` if the update fails. 

I'm not as familiar with how `stripCloneAttrs` works. Would it make more sense to calculate `changed/previousAttrs` when `@load` is called or during `@save`?  And can you still calculate `changedAttrs` after `@load` is called? (i.e. is the old data still available to compare the new attributes to?)

**Comment by adambiggs** [comment]
> I'm not as familiar with how `stripCloneAttrs` works.

It removes `own` attributes from a record clone after the clone is saved. This means next time you access one of those attributes on the clone, they will come from it's prototype in model storage. So if an attribute is changed later on a different clone, the first clone will stay in sync with the data in model storage.

Essentially, the way Spine accesses model storage via `prototype` on clone objects is a baked-in way to track dirty changes.

> Would it make more sense to calculate changed/previousAttrs when @load is called or during @save? And can you still calculate changedAttrs after @load is called?

I think the best place would be in `update()`, before `load()` is called on the root record in `@constructor.irecords`.

**Comment by aeischeid** [comment]
bummer about object.observe. At least in my head that was going to be pretty cool...

I guess we should go ahead somehow with tracking the changedAttrs. At a glance I think the stuff @adambiggs is talking about looks good. An additional thought though, because not all backends will support PATCH do we need to adopt this in an optional way - keep current functionality unless an option is enabled on the model? One other advantage this may have is that on large model collections where tracking the changes might be superfluous or to expensive it can be dodged easily.

**Comment by adambiggs** [comment]
@aeischeid yes that's a good point. This should definitely be an opt-in feature.
